designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,308 @@
+
+When clicked, the Run and Debug CodeLenses locally invoke the Lambda handler function they represent. These Lambda handlers are invoked independent of SAM Templates that exist in the users workspace. The following takes place:
+
+-   A temporary SAM Template is produced, which contains one resource that references the Lambda handler
+-   The temporary SAM Application is built
+-   The resource in the temporary SAM Template is invoked, using configurations set with the Configure CodeLens
+-   (If the Debug CodeLens was clicked) The VS Code debugger is attached to the invoked resource
+
+When clicked, the Configure CodeLens opens a (JSON) configuration file that resides in the workspace and is managed by the Toolkit. All standalone handlers within a workspace will have their configurations stored in this file. Users have autocompletion support in this file. A rich UI for configuration is not considered at this time, but the door remains open to adding a visual editor in the future based on user feedback.
+
+These CodeLenses do not support API Gateway style invokes.
+
+Some users may find CodeLenses within code files distracting, particularly if they are using the Toolkit for features not related to local debugging. Toolkit settings can be used to enable and disable CodeLenses. CodeLenses only appear for languages/runtimes that the Toolkit has implemented Debug support for.
+
+### Graphical User Interface
+
+SAM Template Resources invoked by API Gateway can be locally tested with a REST Workbench.
+
+The REST Workbench is accessed with the command "Open Local REST Workbench for SAM Applications", opens a webview document.
+
+In the REST Workbench, users select a SAM Application Template, build and launch the application (with or without debug mode), and proceed to make a series of REST requests against the running application.
+
+![Workbench Start-up](workbench-start.png)
+
+Users start by selecting the SAM Application they would like to work with. A dropdown lists all SAM Template files detected within the workspace. Next, they launch the application (Run or Debug), which builds and launches the local web server host. At this point, the SAM Application can no longer be changed (users can open a new workbench in order to operate against a different SAM Application).
+
+![Workbench Request Iteration](workbench-request.png)
+
+The running instance of the application is used for all REST queries made by the user. Users define REST parameters (endpoint, request type (GET, POST, etc), a query string, and an optional request body), and press Submit. If the application is run in Debug mode, the Toolkit will attach a debugger after sending the request. The following information is sent to the Toolkit's Output Channel with each REST request:
+
+-   request
+-   response
+-   response code
+-   SAM CLI output
+
+After making code changes, users can click Restart to rebuild the application and start it again.
```

Comment:
It seems weird having the restart button at the top to me, I'd almost have this as a tick-box next to the Submit button.

Also "Submit" as a button name sounds weird to me for doing an HTTP invocation of a REST API.

Reply:
I initially had "Post", but didn't want to collide with the method type POST. Thoughts around "Send"?

I like the idea of restart as a tick box. No point in restarting if you aren't sending a request.